### PR TITLE
Build computational graph manually and not use backward with FunctionHook

### PR DESCRIPTION
**Pull request by disktnk**
~WIP~

Currently ONNX-Chainer uses `FunctionHook`'s backward hook to make computational graph, this cause a bit difficulty to investigate errors and implicit graph construction. This PR will introduce manually making a computational graph. Along with moving convert logic from hook to making graph, this PR also provides clearer code on exporting.

### CODE REVIEW

File: onnx_chainer/graph.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+
+import chainer
+
+from onnx_chainer.functions.converter import FunctionConverterParams
+from onnx_chainer import onnx_helper
+
+
+class Node(object):
+    def __init__(self, node, no=None):
+        self.no = no
```
Comment (shinh) [inline onnx_chainer/graph.py]: How about `order`?

File: onnx_chainer/graph.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            elif isinstance(cand, chainer.FunctionNode):
+                for input_ in cand.inputs:
+                    if input_ is not cand and (input_, cand) not in seen_edges:
+                        add_cand(input_)
+                        seen_edges.add((input_, cand))
+                        input_node = create_node(input_)
+                        cand_node = create_node(cand)
+                        cand_node.add_in_edge(input_node)
+                        input_node.add_out_edge(cand_node)
+                        nodes.add(input_node)
```
Comment (shinh) [inline onnx_chainer/graph.py]: Add `else: assert False`?

File: onnx_chainer/export.py [hunk 3]
```diff
@@ -400,12 +272,12 @@ def _export(model, args, filename, export_params, graph_name, save_text,
...
-    # Backward computation to construct graph
-    with ONNXExport(context, converters, opset_version, network_outputs) as o:
-        chainer.grad(flat_outputs, list(model.params()) + flat_args)
+
+    o = Graph(context, converters, opset_version, network_outputs)
+    o.to_onnx_graph()
 
     implicit_input_names = set(o.inputs.keys()) - param_names -\
-        set(network_inputs.keys())
+        set(network_inputs.keys()) - set(network_outputs.keys())
```
Comment (shinh) [inline onnx_chainer/export.py]: Could you elaborate why this change is necessary?

File: onnx_chainer/graph.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        while cands:
+            _, _, cand = heapq.heappop(cands)
+            if not isinstance(cand, chainer.variable.VariableNode):
+                raise NotImplementedError(
+                    'ONNX-Chainer does not support node type {}'.format(
+                        type(cand)))
+            creator = cand.creator_node
+            if creator is None:
+                continue
+            if (creator, cand) in seen_edges:
```
Comment (shinh) [inline onnx_chainer/graph.py]: Do we really need `seen_edges`? I think we won't visit the same FunctionNode multiple times thanks to the check for `function_nodes` even without this.

File: onnx_chainer/graph.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+                continue
+            assert isinstance(creator, chainer.FunctionNode)
+            creator_id = id(creator)
+            if creator_id in function_nodes:
+                continue
+            seen_edges.add((creator, cand))
+            function_nodes[creator_id] = creator
+
+            for input_ in creator.inputs:
+                if input_ is creator:
```
Comment (shinh) [inline onnx_chainer/graph.py]: Wow. Does this really happen? Could you add a comment for what kind of case this check is necessary?

File: onnx_chainer/graph.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+                    continue
+                if (input_, creator) in seen_edges:
```
Comment (shinh) [inline onnx_chainer/graph.py]: I'd remove this check even if we actually need `seen_edges`. In general, IMO it's a good idea to do dup check only after popping a queue when you do BFS/A* and performance doesn't matter

File: onnx_chainer/graph.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+                else:
+                    output_name = self.context.get_name(output_ref())
+            output_names.append(output_name)
+
+        nodes = self.create_node(
+            func_name, function, input_names, output_names,
+            self.additional_parameters)
+        self.converted_nodes[temp_node_name] = nodes
+
+    def after_rename(self):
```
Comment (shinh) [inline onnx_chainer/graph.py]: I think `rename_outputs` would be a better name. I guess by "after" you mean this is a postprocessing path, but `after_rename` sounds a bit strange. I think "rename" would imply we are in postprocessing.

**Comment by disktnk** [comment]
ChainerCI, test this please
